src/stratis_cli/_errors.py

```diff
@@ -30,6 +32,87 @@ class StratisCliRuntimeError(StratisCliError):
     """
 
 
+class StratisNoChangeError(StratisCliError):
+    """
+    Raised if there was a failure due to no changed state in stratisd's engine.
+    """
+
+    def __init__(self, command, resources):
+        """ Initializer.
+
+            :param str command: the executed command
+            :param str or list of str resources: the target resources
+        """
+        # pylint: disable=super-init-not-called
+        self.command = command
+        self.resources = resources
+
+    def __str__(self):
+        return "The command '{}' has already been completed for resource '{}'".format(
+            self.command, self.resources
+        )
+
+
+class StratisPartialChangeError(StratisCliError):
+    """
+    Raised if there was a failure due to partially changed state in stratisd's engine.
+    """
+
+    def __init__(self, command, changed_resources, unchanged_resources):
+        """ Initializer.
+
+            :param str command: the command run that caused the error
+            :param list of str changed_resources: the target resources that would change
+            :param list of str unchanged_resources: the target resources that would not change
+        """
+        # pylint: disable=super-init-not-called
+        self.command = command
+        self.changed_resources = changed_resources
+        self.unchanged_resources = unchanged_resources
+
+    def __str__(self):
+        msg = "The command '{}' has already been completed for resources {} ".format(
+            self.command, self.unchanged_resources
+        )
+        if self.changed_resources != []:
+            msg += "but not yet for resources {}".format(self.changed_resources)
+        return msg
+
+
+class StratisInUseError(StratisCliError):
+    """
+    Raised if a block device is added as both a blockdev and a cachedev
+    """
+
+    def __init__(self, blockdevs, added_as="cache"):
+        """ Initializer.
+
+            :param list of str blockdevs: the blockdevs that would be added in both tiers
```

Comment:
I have always believed that if you want a complicated type you have to do it separately, like:

```
:param blockdevs: ...
:type blockdevs: list of str
```
if this isn't true anymore, that would be nice, too.

Reply:
Uh oh! I lifted that syntax from somewhere else in the code so the doc string issue may be a little more complicated than originally thought if that's not valid syntax. I'll dig into it a bit more.

Reply:
The documentation for type hints does seem to indicate that type can go [after `:param` and before `name_of_parameter:`.](https://github.com/agronholm/sphinx-autodoc-typehints/blob/master/README.rst#sphinx-autodoc-typehints)

Reply:
Resolved.

Reply:
I went to the source but I don't think that usage is available. Here: https://www.sphinx-doc.org/en/1.8/usage/restructuredtext/domains.html?highlight=%3Atype#the-python-domain. It seems like:

```
:param <name>: <description>
:type <name>: <the type>
```

is the standard and

```
:param <type> <name>: <description>
```
is a shorthand when the type is simple. Quoting: "It is also possible to combine parameter type and description, if the type is a single word, like this..." from the docs. So, I think that clinches it.